File: android/app/src/main/java/uz/electro/remote/ui/ConnectScreen.kt
package uz.electro.remote.ui

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import uz.electro.remote.CarViewModel
import uz.electro.remote.ConnectPhase
import uz.electro.remote.ConnectStatus
import uz.electro.remote.R
import uz.electro.remote.ui.components.*
import uz.electro.remote.ui.components.BrandLockup
import uz.electro.remote.ui.components.CarArt
import uz.electro.remote.ui.theme.*

/**
 * Экран подключения — первое, что видно при запуске.
 *
 * Приложение не открывается само: T-BOX держит модем в спящем режиме, поэтому
 * по кнопке машину сперва будят через сервер — и только когда голова поднялась
 * и ответила, пускаем внутрь. Иначе первые нажатия уходили бы в пустоту.
 */
@Composable
fun ConnectScreen(vm: CarViewModel, status: ConnectStatus) {
    var showSettings by remember { mutableStateOf(false) }
    var showNews by remember { mutableStateOf(false) }
    val news by vm.news.collectAsState()
    val newsRead by vm.newsRead.collectAsState()
    val unreadNews by vm.unreadNews.collectAsState()

    // Путь к машине один — сервер, поэтому никаких разрешений спрашивать не за
    // что: SEND_SMS убран вместе с запасным каналом.
    fun start() {
        if (vm.wakeConfigured) vm.connect() else showSettings = true
    }

    // Настройки открываются полноэкранно ВМЕСТО экрана подключения. Раньше их
    // рисовали перед Column подключения — и Column перекрывал их сверху, отчего
    // кнопка «Настройки» будто «не реагировала» (экран открывался, но был не
    // виден). Здесь настройки — единственное, что рисуется, пока они открыты.
    if (showSettings) {
        androidx.compose.material3.Surface(
            color = ElectroColors.Background,
            modifier = Modifier.fillMaxSize(),
        ) { SettingsScreen(vm) { showSettings = false } }
        return
    }
    if (showNews) {
        LaunchedEffect(Unit) { vm.loadNews() }
        androidx.compose.material3.Surface(color = ElectroColors.Background, modifier = Modifier.fillMaxSize()) {
            NewsScreen(news, newsRead, onRead = { vm.markNewsRead(it.id) },
                onReadAll = { vm.markAllNewsRead() }, onBack = { showNews = false })
        }
        return
    }

    val chosen by vm.selectedVehicle.collectAsState()
    val support by vm.support.collectAsState()
    var showHelp by remember { mutableStateOf(false) }
    val loggedInForFeedback by vm.loggedIn.collectAsState()
    var showFeedback by remember { mutableStateOf(false) }
    // отзыв — только под аккаунтом: без ключа серверу его не принять
    if (showHelp) HelpDialog(support, onFeedback = if (loggedInForFeedback) ({ showFeedback = true }) else null) { showHelp = false }
    if (showFeedback) FeedbackDialog(
        onSend = { kind, text, files, done -> vm.sendFeedback(kind, text, files, done) },
        onDismiss = { showFeedback = false },
    )

    Column(
        Modifier.fillMaxSize().background(ElectroColors.Background)
            .padding(horizontal = Space.x5).padding(top = Space.x6, bottom = Space.x5),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        // Помощь доступна и до подключения — кнопка в правом верхнем углу.
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
            NewsBell(unreadNews, onClick = { showNews = true })
            Spacer(Modifier.width(Space.x2))
            HelpFab(onClick = { showHelp = true })
        }
        Spacer(Modifier.weight(1f))

        // тот же лозунг, что в шапке главного экрана: марка мелко, модель крупно
        BrandLockup(markSize = 26.dp)
        Spacer(Modifier.height(Space.x2))
        // Модель — выбранной машины: с несколькими машинами в парке подпись
        // «C16» над C01 вводила бы в заблуждение ровно там, где это опасно.
        Text(
            chosen?.model ?: "C16",
            style = ElectroType.Display,
            color = ElectroColors.TextPrimary,
        )
        chosen?.name?.takeIf { it.isNotBlank() }?.let {
            Spacer(Modifier.height(Space.x1))
            Text(it, style = ElectroType.Body, color = ElectroColors.TextMuted)
        }
        Spacer(Modifier.height(Space.x6))

        // Рендер — студийная вырезка с прозрачным фоном по модели и цвету кузова,
        // лежит на обычной карточке темы: никакой «вшитой» тёмной подложки.
        val paint by vm.paint.collectAsState()
        Surface(
            color = ElectroColors.SurfaceRaised,
            shape = Radius.Lg,
            modifier = Modifier.fillMaxWidth(),
        ) {
            Box {
                Image(
                    painterResource(CarArt.image(chosen?.model, paint)), null,
                    modifier = Modifier.fillMaxWidth().height(200.dp).padding(horizontal = Space.x2, vertical = Space.x2),
                    contentScale = ContentScale.Fit,
                )
                // статус подключения плашкой в углу hero — как на панели машины
                Box(Modifier.fillMaxWidth().padding(Space.x3), contentAlignment = Alignment.TopEnd) {
                    PhaseBadge(status)
                }
            }
        }

        Spacer(Modifier.height(Space.x5))
        val h = hint(status, vm.wakeConfigured)
        val failed = status.phase == ConnectPhase.Timeout || status.phase == ConnectPhase.Error
        Text(
            h,
            style = ElectroType.Body,
            color = if (failed) ElectroColors.Danger else ElectroColors.TextSecondary,
            textAlign = TextAlign.Center,
        )

        if (status.phase == ConnectPhase.Waiting) {
            Spacer(Modifier.height(Space.x4))
            LinearProgressIndicator(
                Modifier.fillMaxWidth().clip(Radius.Pill),
                color = ElectroColors.Accent,
                trackColor = ElectroColors.SurfaceElevated,
            )
        }

        Spacer(Modifier.weight(1f))

        when (status.phase) {
            ConnectPhase.Idle -> {
                ElectroButton("Подключиться", Modifier.fillMaxWidth()) { start() }
                Spacer(Modifier.height(Space.x2))
                ElectroButton(
                    "Найти машину", Modifier.fillMaxWidth(),
                    style = ButtonStyle.Secondary,
                ) { vm.findCar() }
            }

            ConnectPhase.Sending, ConnectPhase.Waiting -> {
                ElectroButton("Подключение", Modifier.fillMaxWidth(), loading = true) {}
            }

            ConnectPhase.Timeout -> {
                ElectroButton("Разбудить ещё раз", Modifier.fillMaxWidth()) { start() }
                Spacer(Modifier.height(Space.x2))
                ElectroButton(
                    "Всё равно открыть", Modifier.fillMaxWidth(),
                    style = ButtonStyle.Secondary,
                ) { vm.enterAnyway() }
            }

            ConnectPhase.Error -> {
                ElectroButton("Повторить", Modifier.fillMaxWidth()) { start() }
                Spacer(Modifier.height(Space.x2))
                ElectroButton(
                    "Всё равно открыть", Modifier.fillMaxWidth(),
                    style = ButtonStyle.Ghost,
                ) { vm.enterAnyway() }
            }

            ConnectPhase.Connected -> Unit
        }

        Spacer(Modifier.height(Space.x3))
        ElectroButton(
            if (vm.wakeConfigured) "Настройки" else "Настроить подключение",
            Modifier.fillMaxWidth(),
            style = ButtonStyle.Ghost,
        ) { showSettings = true }
        Spacer(Modifier.height(Space.x3))
        Text(appVersion(), style = ElectroType.Caption, color = ElectroColors.TextMuted)
        Spacer(Modifier.height(Space.x3))
    }
}

@Composable
private fun PhaseBadge(status: ConnectStatus) {
    when (status.phase) {
        ConnectPhase.Idle -> StatusBadge(BadgeKind.Offline, "СПИТ")
        ConnectPhase.Sending, ConnectPhase.Waiting -> StatusBadge(BadgeKind.Info, "ПОДКЛЮЧЕНИЕ")
        ConnectPhase.Timeout -> StatusBadge(BadgeKind.Failed, "ОШИБКА")
        ConnectPhase.Error -> StatusBadge(BadgeKind.Failed, "ОШИБКА")
        ConnectPhase.Connected -> StatusBadge(BadgeKind.Online, "НА СВЯЗИ")
    }
}

/** Коротко: без объяснений, что происходит под капотом — только статус. */
private fun hint(status: ConnectStatus, configured: Boolean): String = when {
    !configured && status.phase == ConnectPhase.Idle -> "Войдите на сервер в настройках."
    status.phase == ConnectPhase.Sending || status.phase == ConnectPhase.Waiting -> "Подключение…"
    status.phase == ConnectPhase.Timeout || status.phase == ConnectPhase.Error -> "Ошибка: подключение не удалось"
    else -> ""
}

File: ElectroRemote/UI/Screens/ConnectScreen.swift
import SwiftUI

/// Экран подключения — первое, что видно при запуске. По кнопке машину будят
/// через сервер, и только когда голова ответила, пускают внутрь.
struct ConnectScreen: View {
    @Environment(\.palette) private var p
    @ObservedObject var vm: CarViewModel
    let status: ConnectStatus

    @State private var showSettings = false
    @State private var showHelp = false
    @State private var showNews = false
    @State private var showShop = false

    private func start() {
        if vm.wakeConfigured { vm.connect() } else { showSettings = true }
    }

    var body: some View {
        if showSettings {
            SettingsScreen(vm: vm) { showSettings = false }
        } else if showShop {
            ShopScreen(products: vm.products, loggedIn: true, model: vm.selectedVehicle?.model,
                                   onOrder: { id, qty, phone, comment, done in vm.order(id, qty: qty, phone: phone, comment: comment, done: done) },
                                   onBack: { showShop = false }, onRefresh: { vm.loadProducts() })
        } else if showNews {
            NewsScreen(items: vm.news, isRead: { vm.isRead($0) }, onRead: { vm.markRead($0) },
                       onReadAll: { vm.markAllRead() }, onBack: { showNews = false }, onRefresh: { vm.loadNews() })
        } else {
            content
        }
    }

    private var content: some View {
        let chosen = vm.selectedVehicle
        return VStack(spacing: 0) {
            HStack(spacing: Space.x2) {
                LangPicker(compact: true)
                Spacer()
                ShopFab { showShop = true }
                NewsBell(unread: vm.unreadNews) { showNews = true }
                HelpFab { showHelp = true }
            }
            Spacer()

            BrandLockup(markSize: 26)
            Spacer().frame(height: Space.x2)
            Text(chosen?.model ?? "C16").font(ElectroType.display).foregroundStyle(p.textPrimary)
            if let name = chosen?.name, !name.isEmpty {
                Spacer().frame(height: Space.x1)
                Text(name).font(ElectroType.body).foregroundStyle(p.textMuted)
            }
            Spacer().frame(height: Space.x6)

            // Рендер — студийная вырезка по модели и цвету кузова, на карточке темы.
            ZStack(alignment: .topTrailing) {
                Image(CarArt.imageName(chosen?.model, vm.paint)).resizable().scaledToFit()
                    .frame(maxWidth: .infinity).frame(height: 200)
                    .padding(.horizontal, Space.x2).padding(.vertical, Space.x2)
                PhaseBadge(status: status).padding(Space.x3)
            }
            .frame(maxWidth: .infinity)
            .background(p.surfaceRaised)
            .clipShape(RoundedRectangle(cornerRadius: Radius.lg, style: .continuous))

            Spacer().frame(height: Space.x5)
            let failed = status.phase == .timeout || status.phase == .error
            Text(hint(status, vm.wakeConfigured)).font(ElectroType.body).foregroundStyle(failed ? p.danger : p.textSecondary)
                .multilineTextAlignment(.center)

            if status.phase == .waiting {
                Spacer().frame(height: Space.x4)
                ProgressView().progressViewStyle(.linear).tint(p.accent)
            }

            Spacer()

            VStack(spacing: Space.x2) {
                switch status.phase {
                case .idle:
                    ElectroButton(text: L("Подключиться")) { start() }
                    ElectroButton(text: L("Найти машину"), style: .secondary, enabled: !vm.busy) { vm.findCar() }
                case .sending, .waiting:
                    ElectroButton(text: L("Подключение"), loading: true) {}
                case .timeout:
                    ElectroButton(text: L("Разбудить ещё раз")) { start() }
                    ElectroButton(text: L("Всё равно открыть"), style: .secondary) { vm.enterAnyway() }
                case .error:
                    ElectroButton(text: L("Повторить")) { start() }
                    ElectroButton(text: L("Всё равно открыть"), style: .ghost) { vm.enterAnyway() }
                case .connected:
                    EmptyView()
                }
            }

            Spacer().frame(height: Space.x3)
            ElectroButton(text: vm.wakeConfigured ? L("Настройки") : L("Настроить подключение"), style: .ghost) { showSettings = true }
            Spacer().frame(height: Space.x3)
            Text(appVersion()).font(ElectroType.caption).foregroundStyle(p.textMuted)
            Spacer().frame(height: Space.x3)
        }
        .padding(.horizontal, Space.x5)
        .padding(.top, Space.x6)
        .padding(.bottom, Space.x5)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(p.background)
        .sheet(isPresented: $showHelp) { HelpSheet(support: vm.support, feedbackVM: vm.loggedIn ? vm : nil) }
        .onAppear { CarModels.prefetch(vm.selectedVehicle?.model) }   // 3D-модель для главной — заранее
    }
}

private struct PhaseBadge: View {
    let status: ConnectStatus
    var body: some View {
        switch status.phase {
        case .idle: StatusBadge(kind: .offline, text: L("СПИТ"))
        case .sending, .waiting: StatusBadge(kind: .info, text: L("ПОДКЛЮЧЕНИЕ"))
        case .timeout: StatusBadge(kind: .failed, text: L("ОШИБКА"))
        case .error: StatusBadge(kind: .failed, text: L("ОШИБКА"))
        case .connected: StatusBadge(kind: .online, text: L("НА СВЯЗИ"))
        }
    }
}

/// Коротко: без объяснений, что происходит под капотом — только статус.
private func hint(_ status: ConnectStatus, _ configured: Bool) -> String {
    if !configured && status.phase == .idle { return L("Войдите на сервер в настройках.") }
    switch status.phase {
    case .sending, .waiting: return L("Подключение…")
    case .timeout, .error: return L("Ошибка: подключение не удалось")
    case .idle, .connected: return ""
    }
}
